pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -191,30 +191,82 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 		gc.workerLock.Lock()
 		defer gc.workerLock.Unlock()
 
-		// Resetting the REST mapper will also invalidate the underlying discovery
-		// client. This is a leaky abstraction and assumes behavior about the REST
-		// mapper, but we'll deal with it for now.
-		gc.restMapper.Reset()
-
-		// Perform the monitor resync and wait for controllers to report cache sync.
-		//
-		// NOTE: It's possible that newResources will diverge from the resources
-		// discovered by restMapper during the call to Reset, since they are
-		// distinct discovery clients invalidated at different times. For example,
-		// newResources may contain resources not returned in the restMapper's
-		// discovery call if the resources appeared in-between the calls. In that
-		// case, the restMapper will fail to map some of newResources until the next
-		// sync period.
-		if err := gc.resyncMonitors(newResources); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
-			return
-		}
-		// TODO: WaitForCacheSync can block forever during normal operation. Could
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
-			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
-			return
+		// Once we get here, we should not unpause workers until we've successfully synced again
+		for attempt := 0; ; attempt++ {
+			// If this is a re-attempt, check if available resources have changed
+			if attempt > 0 {
+				newResources = GetDeletableResources(discoveryClient)
```

Comment:
We may want to check if the list is empty again, like here https://github.com/kubernetes/kubernetes/pull/64235/files#diff-1481f32cd1a60ce2a66e5d8b5e6cf383R175 but instead of returning we should probably continue,

if GetDeletableResources encounters any errors it will return an empty list of resources. if we allow it to go forward with the sync with an empty list of resources, resyncMonitors will remove all the work that has been done.

Reply:
Good idea. Probably add a protection against a hotloop here and in the resyncFailure case as well

Reply:
checked for zero length rediscovery results, used `wait.PollImmediateInfinite` instead of a plain for loop to get hotloop and panic protection